packages/pyright-internal/src/analyzer/program.ts

```diff
@@ -2032,6 +2032,10 @@ export class Program {
             }
 
             // We might not have the file info if binding failed for whatever reasons.
+            if (!AnalyzerNodeInfo.hasAnalyzerInfo(parseResults.parseTree)) {
```

Comment:
Do we understand the root cause of this problem? I don't understand why a parse tree would ever _not_ have a fileInfo in its root node. This looks like it's just masking a deeper problem.

Reply:
this is only place we ever set `fileInfo` - https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/analyzer/sourceFile.ts#L816

so, one case would be if `binding` failed. 

or this call (https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/analyzer/program.ts#L2019) return without actually calling `binding`

Reply:
this crash is from `telemetry` so unfortunately we don't have repro steps.

Reply:
> so, one case would be if binding failed.

This would be a problem only if we catch an exception very early in the `bind` method, before we even construct and run the `Binder`. I think that's really unlikely unless there's some bad data structure corruption issue, in which case we're in a pretty bad situation and will likely crash anyway.

Reply:
the other case is this

> this call (https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/analyzer/program.ts#L2019) return without actually calling binding

basically, we call `_checkTypes` and assume the file would have bound. but in certain case, due to this, it won't.

https://github.com/microsoft/pyright/blob/e8f0965e5b0429232af0c8881781f0b5045b404e/packages/pyright-internal/src/analyzer/program.ts#L1899-L1912

For example, `isCheckingRequired` would return `false` even if `dump cache` dumped binding info (basically parse tree). and `binder` has not been called yet but parser did.

another approach to fix it would be not attaching data to parse node so they doesn't go away when parse tree go away like how data produced by checker do.

or force `_checkTypes` to run if `binding` information is not there.

Reply:
In that case, the file will not be marked as bound (i.e. `isBindingNeeded` will still be true). It sounds like we should check whether the file has been bound. This is a much cleaner and less fragile approach than checking for certain side effects of binding.

```python
            if (file.sourceFile.isBindingRequired()) {
                continue;
            }
```

Reply:
ya, that will work as well.